clusterloader2/pkg/measurement/common/slos/job_startup_latency.go

```diff
@@ -0,0 +1,244 @@
+	event := &eventData{obj: obj, recvTime: time.Now()}
+	p.eventQueue.Add(event)
+}
+
+func (p *jobStartupLatencyMeasurement) processEvents() {
+	for p.processNextWorkItem() {
+	}
+}
+
+func (p *jobStartupLatencyMeasurement) processNextWorkItem() bool {
+	item, quit := p.eventQueue.Get()
+	if quit {
+		return false
+	}
+	defer p.eventQueue.Done(item)
+	event, ok := item.(*eventData)
+	if !ok {
+		klog.Warningf("Couldn't convert work item to evetData: %v", item)
+		return true
+	}
+	p.processEvent(event)
+	return true
+}
+
+// processEvent processes job state change events:
+// uses Phase Latency utility to record job state transitions
+// it currently captures the following transitions:
+// JobCreated (job.CreationTimestamp.Time) -> JobStarted (job.Status.StartTime.Time)
+// JobStarted (job.Status.StartTime.Time) -> JobCompleted (job.Status.CompletionTime.Time)
+func (p *jobStartupLatencyMeasurement) processEvent(event *eventData) {
+	obj, _ := event.obj, event.recvTime
+	if obj == nil {
+		return
+	}
+	job, ok := obj.(*batchv1.Job)
+	if !ok {
+		return
+	}
+	key := createMetaNamespaceKey(job.Namespace, job.Name)
+	if _, found := p.jobStateEntries.Get(key, jobCreated); !found {
+		p.jobStateEntries.Set(key, jobCreated, job.CreationTimestamp.Time)
+	}
+	if _, found := p.jobStateEntries.Get(key, jobStarted); !found {
+		if job.Status.StartTime != nil {
+			p.jobStateEntries.Set(key, jobStarted, job.Status.StartTime.Time)
+		}
+	}
+	if _, found := p.jobStateEntries.Get(key, jobCompleted); !found {
+		if job.Status.CompletionTime != nil {
+			p.jobStateEntries.Set(key, jobCompleted, job.Status.CompletionTime.Time)
+		}
+	}
+}
+
+func (p *jobStartupLatencyMeasurement) stop() {
+	if p.isRunning {
+		p.isRunning = false
+		close(p.stopCh)
+		p.eventQueue.ShutDown()
+	}
+}
+
+var jobStartupTransitions = map[string]measurementutil.Transition{
+	"create_to_start": {
+		From: jobCreated,
+		To:   jobStarted,
+	},
+	"start_to_complete": {
+		From: jobStarted,
+		To:   jobCompleted,
+	},
+}
+
+type jobStartupLatencyCheck struct {
+	namePrefix string
+	filter     measurementutil.KeyFilterFunc
+}
+
+// gather collects job startup latency and calculates percentiles using Phase Latency utility
+// it waits for all jobs to be completed before collecting the metrics or times out
+func (p *jobStartupLatencyMeasurement) gather(c clientset.Interface, identifier string, schedulerName string, timeout time.Duration) ([]measurement.Summary, error) {
+	klog.V(2).Infof("%s: gathering job startup latency measurement...", p)
+	if !p.isRunning {
+		return nil, fmt.Errorf("metric %s has not been started", jobStartupLatencyMeasurementName)
+	}
+	condition := func() (bool, error) {
+		return p.jobStateEntries.Count(jobCreated) == p.jobStateEntries.Count(jobCompleted), nil
+	}
+	if err := wait.Poll(checkCompletedJobsInterval, timeout, condition); err != nil {
+		klog.V(2).Infof("Timed out waiting for all jobs to complete: %v", err)
+	}
+	p.stop()
+	var err error
```

Comment:
I doesn't look like this line is needed.

Reply:
Which one? Do you mean the log output?

Reply:
Ok, got you. The `err` is always `nil` removed